Move header content to a popup

beszel opened the pull request:
Resolves #68, resolves #101, resolves #70, resolves #79, resolves #110 

This may still need unit tests, but I want to put it in review sooner rather than later. All tests are passing.

Everything from the header should be removed, although I might have missed something that was used by the header and is now unused. Is everything as clean as can be? Feel free to nitpick, as I'm always interested in best practices.

beszel committed:
Add popup that redundantly displays header text

beszel committed:
Add popup that redundantly displays header text

beszel committed:
Add popup that redundantly displays header text

beszel committed:
Add popup that redundantly displays header text

beszel committed:
Reproduce all DGtW header functionality in popup

The "Dismiss this header" button is not included in the popup, for
obvious reasons. The "Restore original content" button is fully working.
The message remains the same after restoring content.

beszel committed:
Reproduce all DGtW header functionality in popup

The "Dismiss this header" button is not included in the popup, for
obvious reasons. The "Restore original content" button is fully working.
The message remains the same after restoring content.

beszel committed:
Reproduce all DGtW header functionality in popup

The "Dismiss this header" button is not included in the popup, for
obvious reasons. The "Restore original content" button is fully working.
The message remains the same after restoring content.

beszel committed:
Reproduce all DGtW header functionality in popup

The "Dismiss this header" button is not included in the popup, for
obvious reasons. The "Restore original content" button is fully working.
The message remains the same after restoring content.

beszel committed:
Apply style to popup buttons

beszel committed:
Apply style to popup buttons

beszel committed:
Run full Puppeteer #101

Also fixed "Restore original content" button in popup.

beszel committed:
Run full Puppeteer #101

Also fixed "Restore original content" button in popup.

beszel committed:
Add e2e test for popup restore button

This uses a hardcoded popup URL, which should be changed.

beszel committed:
Add e2e test for popup restore button

This uses a hardcoded popup URL, which should be changed.

beszel committed:
Incomplete e2e tests for popup

beszel committed:
Incomplete e2e tests for popup

beszel committed:
Update tests to use popup

The extension ID is also static now.

beszel committed:
Update tests to use popup

The extension ID is also static now.

beszel committed:
Change "show changes" to a checkbox

With appropriate end-to-end tests

beszel committed:
Change "show changes" to a checkbox

With appropriate end-to-end tests

beszel committed:
Rewrite extension status request

When the popup requests the extension status from the content script,
it now receives a keyword telling it which message to display. It
previously received the entire message that it should display.

The Status enum can't be imported from `popup.js`, so a temporary
solution is used.

beszel committed:
Rewrite extension status request

When the popup requests the extension status from the content script,
it now receives a keyword telling it which message to display. It
previously received the entire message that it should display.

The Status enum can't be imported from `popup.js`, so a temporary
solution is used.

beszel committed:
Replace toggle buttons with checkboxes

beszel committed:
Replace toggle buttons with checkboxes

beszel committed:
Show the restore button when appropriate

beszel committed:
Show the restore button when appropriate

beszel committed:
Fix status enum

beszel committed:
Fix status enum

beszel committed:
Move popup style to `popup.css`

beszel committed:
Move popup style to `popup.css`

beszel committed:
Remove header

beszel committed:
Remove header

beszel commented on src/popup.js:
```diff
@@ -0,0 +1,157 @@
+            statusText += "technical incompatibility.";
+
+            // Show no buttons
+            hideElement(ids.showChanges);
+            hideElement(ids.showHighlights);
+            hideElement(ids.restore);
```
Can this be refactored using a helper method for less repetition?

ProfJanetDavis replied:
Yeah, I see the repetition below. I could imagine adding a function like this:
<pre>
setElementVisibility(showChanges:bool, showHighlights:bool, restore:bool):void
</pre>
That would be a little more concise. But it would also be a little less self-explanatory, since the settings for the three elements would be identified by argument position rather than name. What do you think about the tradeoff?

beszel replied:
I think the method you wrote is the best we could do, and I'm not entirely content with it. Let's leave the three `hideElement` calls for clarity.

ProfJanetDavis replied:
I agree. Thanks for raising the issue.

ProfJanetDavis commented:
It's kind of fun to see the e2e tests running in the browser! But I am having trouble running the unit tests. Is it possible you have uncommitted changes?

<pre>
[2:54pm] janet@brick (~/chrome-extensions/degender-the-web): git checkout popup
Already on 'popup'
Your branch is up to date with 'origin/popup'.
[2:55pm] janet@brick (~/chrome-extensions/degender-the-web): npm run unittest

> degender-the-web@0.8.0 unittest /Users/janet/chrome-extensions/degender-the-web
> mocha-headless-chrome -f http://127.0.0.1:8080/test/unit

[Error: SyntaxError: The requested module '../../src/excluded-domains.js' does not provide an export named 'getWhyExcluded']


  0 passing (1ms)

[2:59pm] janet@brick (~/chrome-extensions/degender-the-web): grep getWhyExcluded test/unit/*
test/unit/excluded-domains.test.js:    getWhyExcluded
test/unit/excluded-domains.test.js:    describe("getWhyExcluded", function() {
test/unit/excluded-domains.test.js:                chai.expect(getWhyExcluded("www." + d)).to.equal(
grep: test/unit/grammar: Is a directory
</pre>

beszel committed:
Remove tests for excluded domain functions

Remove tests for `getExcludedDomain` and `getWhyExcluded`

beszel committed:
Remove tests for excluded domain functions

Remove tests for `getExcludedDomain` and `getWhyExcluded`

beszel commented:
My mistake – I removed a couple of functions for identifying if and why a site isn't supported. Currently the popup just says that the site isn't supported with no site-specific details and no domain name.
I've removed the leftover tests now, but I could instead add back the unused functions and tell the linter not to worry about it.

ProfJanetDavis commented on test/e2e/bootstrap.js:
```diff
@@ -6,24 +6,39 @@ const globalVariables = _.pick(global, [
+    args: [
+        `--disable-extensions-except=${extensionPath}`,
+        `--load-extension=${extensionPath}`
+    ]
```
Nice.

ProfJanetDavis commented on test/e2e/bootstrap.js:
```diff
@@ -6,24 +6,39 @@ const globalVariables = _.pick(global, [
+    global.unsupportedURL = "https://www.facebook.com";
+    global.popupURL =
+        "chrome-extension://kgeehecadkggegiegoamiabpdjpgjkhg/src/popup.html?test=true";
+    // TODO Does "dgtw" need to be specified for IDs in the popup?
```
Probably not. I used "dgtw" for the ids and classes used in the content script to avoid accidentally shadowing ids and classes in the original content (which we have no control over). For the popup, it's a non-issue.

ProfJanetDavis commented on test/e2e/bootstrap.js:
```diff
@@ -6,24 +6,39 @@ const globalVariables = _.pick(global, [
         content: "#text",
-        header: "#dgtw-header",
-        toggle: "#dgtw-toggle",
+        header: "#dgtw-header", // TODO Remove when removing header
```
Before merging the PR, all TODOs should be completed or moved to GitHub issues.

ProfJanetDavis commented on test/e2e/content/case.excluded-domain.test.js:
```diff
@@ -1,10 +1,45 @@
     it("should load the Chrome extension");
-    it("should not change the text");
-    it("should explain in the header");
-    it("should not have a 'Show changes/highlights' button");
```
Thank you so much for taking care of these tests I proposed before but could not implement.

ProfJanetDavis commented on test/e2e/content/case.excluded-domain.test.js:
```diff
@@ -1,10 +1,45 @@
+    let originalContent;
+
+    before(async function() {
+        this.timeout(10000); // Loading Facebook can take more than 8 seconds.
```
This is a helpful comment.

ProfJanetDavis commented:
Hmm. As long as the data file includes the reason for excluding the site, that information should be reflected in the popup. But I included the reason there because of #34, and so far we have not come up with any domains to exclude on such grounds that aren't addressed by looking for "gender" or personal pronoun specifiers. 

So, what do you think: Should we remove the reasons from `data/excluded-domains.js`, making `excludedDomains` a list rather than a dictionary? Or should we show the given reason in the popup?

ProfJanetDavis commented on src/popup.html:
```diff
@@ -0,0 +1,24 @@
+    <div>
+        <button id="dgtw-restore" class="hide">
+            Restore original content
+        </button>
```
I'm playing with the popup on [this news story](https://www.chicagotribune.com/entertainment/ct-lilly-wachowski-glaad-media-awards-20160405-story.html) from the manual tests, and I'm finding the interaction between the button and the popup confusing. Why? Because after you restore the original content, the checkbox has no effect.

It seems like when you restore the original page content, the popup content needs to change too. It should probably say that the original content has been restored. You can't show/hide changes/highlights, so that option should be removed. You can't restore the original content again, so that option should be removed too. The only possible option to provide is to refresh the page so the content script runs again. Does this make sense?

beszel commented:
I forgot about #34 when I saw that sites were only excluded due to technical incompatibility. Let me see what I can do to account for it in the popup.

ProfJanetDavis commented on src/popup.js:
```diff
@@ -0,0 +1,157 @@
+            break;
+
+        case Status.restoredOriginal:
+            // TODO This is not yet used.
```
See my comments on `popup.html`. We seem to be thinking alike here.

ProfJanetDavis commented on src/status.js:
```diff
@@ -0,0 +1,8 @@
+    mentionsGender: "mentionsGender",
+    replacedPronouns: "replacedPronouns",
+    noGenderedPronouns: "noGenderedPronouns"
+};
```
It seems like a waste of a dictionary to have the values identical to the keys. I wonder  it would make sense to put the status messages here, as the dictionary values, rather than in the `popup.js` code.

beszel replied:
I think it makes sense for the popup to be responsible for the status messages it displays. The popup needs to get a status (like 'mentionsGender') to tell it which interface items to show or hide. It would be more cumbersome to try and match a long string, and unnecessarily obtuse to use numerical status codes.

ProfJanetDavis replied:
Right, I think I'm getting muddled by JavaScript's object/dictionary duality. I'm willing to leave responsibility for the messages in the popup (at least for now).

ProfJanetDavis reviewed (changes requested):
It's awesome to see so much code deleted! Woohoo! 

I noted several things to improve, including a few that you seemed to consider incomplete yourself. But I'm so excited to see how it's coming along overall, and I'm pleased to see how you've been updating the tests along with the code.

ProfJanetDavis commented on manifest.json:
```diff
@@ -4,6 +4,11 @@
     "short_name": "DGtW",
     "description": "'Degender the Web' replaces gendered pronouns on most Web pages with 'they/them/their'.",
     "version": "0.8.0",
```
Please bump the version number to 0.9.0. This should also be done in package.json and package-lock.json.

beszel committed:
Revert "Remove tests for excluded domain functions"

This reverts commit 2694fc24397064ea67772958dff40e5761fecd27.

The removed tests should be left in, as the functions they are testing
will be put back in. Tests will not pass on this commit.

beszel committed:
Revert "Remove tests for excluded domain functions"

This reverts commit 2694fc24397064ea67772958dff40e5761fecd27.

The removed tests should be left in, as the functions they are testing
will be put back in. Tests will not pass on this commit.

beszel committed:
Show reason that extension is disabled on a site

If the extension had to be disabled on a particular site, the header
would explain why. Now the popup explains why.

beszel committed:
Show reason that extension is disabled on a site

If the extension had to be disabled on a particular site, the header
would explain why. Now the popup explains why.

beszel committed:
Explain in popup when the content is restored

A broken test was found and fixed as part of this commit.

beszel committed:
Explain in popup when the content is restored

A broken test was found and fixed as part of this commit.

beszel committed:
Clean up IDs and selectors

beszel committed:
Clean up IDs and selectors

beszel committed:
Update version number and remove TODOs

beszel committed:
Update version number and remove TODOs

ProfJanetDavis commented on src/popup.html:
```diff
@@ -0,0 +1,24 @@
+  </head>
+
+  <body>
+    <div id="dgtw-status"></div>
```
One tiny change - I'd like to see some spacing between the status message and the controls. You may be able to accomplish this satisfactorily just by using `<p>` rather than `<div>`. If that's not satisfactory, there's always the CSS.

ProfJanetDavis reviewed (approved):
Thanks for addressing my feedback. I asked for one small style change that I trust you to accomplish in a reasonable way. Go ahead and merge when that's done!

ProfJanetDavis commented:
Whoops - just realized you were not done yet, so approval is premature.

beszel committed:
Add space between status message and controls

beszel committed:
Add space between status message and controls

beszel committed:
Merge branch 'popup' of github.com:glam-lab/degender-the-web into popup

beszel committed:
Add reload button to un-restore content

beszel committed:
Make end-to-end tests more consistent and thorough

beszel committed:
Remove leftover header CSS
